### PR TITLE
Hide the Product by attribute, Product by tag, and Product by category templates from the list on the Site Editor

### LINKED ISSUE #42531: Hide the `Products by Category, Tag, and Attribute` templates from the templates list
Hide those three templates from the list of templates by default. Instead, they should appear in the `Add New` button menu. However, if the theme or the user has customized one of those three templates, it will appear in the templates list. (The tl;dr is that they should behave as the Category and Tag templates from WP core.)

### PR DESCRIPTION
**Pull request by albarin**
The goal of this PR is to hide the `Product by attribute`, `Product by tag`, and `Product by category` templates from the list on the Site Editor if they are not customized by the user (stored on the DB) or the theme (template on the theme folder). They don't need to appear since they all fall back to the `Product Catalog`.
Instead, they should appear in the `Add New` button menu. `Product by tag` and `Product by category`, already appear on the menu, this PR also extends the default templates (using the `default_template_types` filter) to make the `Product by attribute` button available in Gutenberg. 
To test the whole flow, [this PR](https://github.com/WordPress/gutenberg/pull/46270) in Gutenberg is also needed since the `Add new` menu is controlled from there.

Fixes https://github.com/woocommerce/woocommerce/issues/42531

### Screenshots

| Before (`Products by Category, Tag, and Attribute` show in the list) | After (only `Product Catalog` shows since it's the fallback template for the other three) |
| ------ | ----- |
| Screenshot 2022-11-29 at 14 09 29 | Screenshot 2022-11-29 at 14 09 54 |
| Screenshot 2022-11-30 at 12 12 36 | Screenshot 2022-11-30 at 12 12 07 |

#### Pre-requisites for testing
- Make sure you are on [this branch](https://github.com/WordPress/gutenberg/pull/46125) on Gutenberg.
- Make sure you have at least one tag and some products tagged with it.
- Make sure you have at least one Product Attribute with `Enabled Archives` and some products associated with it.
- Make sure you have at least one category and some products on it.
- Make sure you are using the "Twenty Twenty Two" theme (or a blocks theme).
- Make sure to remove all template customizations from Site Editor > Browse all templates.

#### `Product by category/tag/attribute` fall back to `Product Catalog` and don't show up in the list
1. Go to the Site Editor > Browse all templates (`wp-admin/site-editor.php?postType=wp_template`)
2. Make sure the `Product by attribute`, `Product by tag`, and `Product by category` templates don't appear on the list of templates.
3. Edit the `Product Catalog` template, add some customization and save.
4. On the front end, go to `/shop` and make sure you see the customization you just did.
5. On the front end, go also to a category, a tag, and an attribute page (like `product-category/clothing/`, `/product-tag/music`, and `color/red`, depending on your store configuration).
6. Check that all 3 pages are using the `Product Catalog` template, meaning you see exactly the same customization you did on the `Product Catalog`.

#### Add customization to the `Product by attribute` template / Clear customization
1. Go to the Site Editor > Browse all templates (`wp-admin/site-editor.php?postType=wp_template`)
2. Click the `Add new` button and click on the `Product by Attribute` menu item.
3. On the `Product by Attribute` template page, add some customization and save.
4. Go back to the Site Editor template list (`wp-admin/site-editor.php?postType=wp_template`).
5. Make sure now you see the `Product by attribute` template on the list.
6. On the front end, go to an attribute page (like `color/red`, depending on your store configuration) and make sure you see the customization you just did in step 3.
7. On the front end, go to `/shop` and make sure you see the customization you did on the `Product Catalog`.
8. On the front end, go also to a category and a tag (like `product-category/clothing/`, `/product-tag/music`, depending on your store configuration) and make sure also you see the customization you did on the `Product Catalog`.
9. Go to the Site Editor > Browse all templates (`wp-admin/site-editor.php?postType=wp_template`)
10. Delete the `Product by attribute` customization.
11. On the front end, go to an attribute page (like `color/red`, depending on your store configuration) and make sure you now see the `Product Catalog` customization (it should fall back to it again).
12. Repeat the steps above with `Product by tag` and `Product by catalog`

#### `Product by category/tag/attribute` from theme take priority over `Product Catalog`
1. Make you have no customizations on any template on `/wp-admin/site-editor.php?postType=wp_template`.
2. Add a file template called `archive-product.html` to the `templates` folder of your theme with some identifiable changes.
3. Add a file template called `taxonomy-product_attribute.html` to your `templates` folder of your theme with some identifiable changes.
4. On the Site Editor (`/wp-admin/site-editor.php?postType=wp_template`) check the `Product by Attribute` appears on the template list and the `Added by` column shows the name of your theme.
5. On the front end, go to a product attribute page and make sure it's using the `taxonomy-product_attribute.html` template.
6. On the front end, go to `/shop`, to a category, and a tag page and check that they are using the `archive-product.html`.
7. Repeat for `Product by Tag` (file `taxonomy-product_tag.html`) and `Product by Category` (file `taxonomy-product_cat.html`).

### WooCommerce Visibility
* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Changelog

> Hide the `Products by Category, Tag, and Attribute` templates from the templates list if they are not customized by the user or the theme.

### CODE REVIEW

File: src/BlockTemplatesController.php [hunk 1]
```diff
@@ -64,15 +64,81 @@ protected function init() {
...
+
+			// We call again the `get_block_templates` function to carry on with the normal execution but with the
+			// modified hierarchy.
+			$block_templates = get_block_templates( $query, $template_type );
+			return array_filter(
+				$block_templates,
+				function( $template ) use ( $query ) {
+					if ( 'archive-product' === $template->slug ) {
+						$template->slug = $query['slug__in'][0];
+						return $template;
```
Comment (thealexandrelara) [inline src/BlockTemplatesController.php]: I saw that this variable is already being returned on line 118, so just confirming if we need this early return here or if we can leave this condition only to modify the slug.

Comment (albarin) [reply src/BlockTemplatesController.php]: Yes, you are right, it's not needed. Fixed! 🙌

**Review by thealexandrelara — APPROVED** [review]
LGTM!

**Review by thealexandrelara — APPROVED** [review]
LGTM!

**Comment by nerrad** [comment]
Let's make sure this PR gracefully degrades if the relevant new code in GB isn't available. That allows the PR to be merged sooner without worry of breaking anything.

**Pull request closed without merging**